### PR TITLE
SYCL parallel_for MDRangePolicy

### PR DESCRIPTION
**Pull request by masterleinad**
Needs #3582. ~~So far only tested with the incremental test.~~

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
This should be ready.

**Comment by masterleinad** [comment]
Only one of the HIP CI checks is timing out which I am pretty sure is unrelated.

File: core/unit_test/TestMDRange.hpp [hunk 1]
```diff
@@ -1411,8 +1411,13 @@ struct TestMDRange_3D {
...
       using point_type = typename range_type::point_type;
 
+#ifndef KOKKOS_ENABLE_SYCL
       range_type range(point_type{{0, 0, 0}}, point_type{{N0, N1, N2}},
-                       tile_type{{8, 8, 8}});
+                       til = e_type{{8, 8, 8}});
+#else
+      range_type range(point_type{{0, 0, 0}}, point_type{{N0, N1, N2}},
+                       tile_type{{8, 8, 4}});
+#endif
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/TestMDRange.hpp:1420]: Yes.

File: core/unit_test/TestRange.hpp [hunk 2]
```diff
@@ -190,8 +192,9 @@ struct TestRange {
   KOKKOS_INLINE_FUNCTION
   void operator()(const VerifyOffsetTag &, const int i) const {
     if (i + offset != m_flags(i)) {
+      // FIXME_SYCL printf needs a workaround
 #ifndef __SYCL_DEVICE_ONLY__
-      printf("TestRange::test_for_error at %d != %d\n", i + offset, m_flags(i));
+      printf("TestRange::test_for error at %d != %d\n", i + offset, m_flags(i));
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/TestRange.hpp:197]: No.

File: core/src/KokkosExp_MDRangePolicy.hpp [hunk 3]
```diff
@@ -424,21 +429,38 @@ struct MDRangePolicy : public Kokkos::Impl::PolicyTraits<Properties...> {
...
+                       Kokkos::Experimental::HIP>::value &&
+          m_prod_tile_dims > 1024) {
+        printf(" Tile dimensions exceed HIP limits\n");
+        Kokkos::abort(
+            "HIP ExecSpace Error: MDRange tile dims exceed maximum number "
+            "of threads per block - choose smaller tile dims");
+      }
+#endif
+#if defined(KOKKOS_ENABLE_SYCL)
+      // FIXME_SYCL query the limit instead
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/KokkosExp_MDRangePolicy.hpp:454]: `m_device.get_info<device::max_work_group_size>()` should work.

**Review by dalg24 — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp [hunk 4]
```diff
@@ -130,4 +132,155 @@ class Kokkos::Impl::ParallelFor<FunctorType, ExecPolicy,
...
+  using array_index_type = typename Policy::array_index_type;
+  using index_type       = typename Policy::index_type;
+  using LaunchBounds     = typename Policy::launch_bounds;
+  using WorkTag          = typename Policy::work_tag;
+
+  const FunctorType m_functor;
+  const Policy m_policy;
+
+  ParallelFor()        = delete;
+  ParallelFor& operator=(ParallelFor const&) = delete;
```
Comment (nliber) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:152]: We should declare the copy/move constructor and move assignment operator as deleted and the destructor as defaulted (Rule of 5).

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
Please address https://github.com/kokkos/kokkos/pull/3583#discussion_r530514807
and cleanup history and I will merge

**Comment by masterleinad** [comment]
> Please address [#3583 (comment)](https://github.com/kokkos/kokkos/pull/3583#discussion_r530514807)
> and cleanup history and I will merge

@dalg24 Here you go!

**Comment by masterleinad** [comment]
All CI except for Travis has passed.

**Pull request merged**